Sources/AppcuesKit/Push/PushMonitor.swift

```diff
@@ -0,0 +1,132 @@
+
+internal protocol PushMonitoring: AnyObject {
+    var pushEnabled: Bool { get }
+    var pushBackgroundEnabled: Bool { get }
+    var pushPrimerEligible: Bool { get }
+
+    func refreshPushStatus(completion: ((UNAuthorizationStatus) -> Void)?)
+
+    // Using `userInfo` as a parameter to be able to mock notification data.
+    func didReceiveNotification(userInfo: [AnyHashable: Any], completionHandler: @escaping () -> Void) -> Bool
+}
+
+internal class PushMonitor: PushMonitoring {
+
+    private weak var appcues: Appcues?
+    private let storage: DataStoring
+
+    private var pushAuthorizationStatus: UNAuthorizationStatus = .notDetermined
+
+    var pushEnabled: Bool {
+        pushAuthorizationStatus == .authorized && storage.pushToken != nil
+    }
+
+    var pushBackgroundEnabled: Bool {
+        storage.pushToken != nil
+    }
+
+    var pushPrimerEligible: Bool {
+        pushAuthorizationStatus == .notDetermined
+    }
+
+    init(container: DIContainer) {
+        self.appcues = container.owner
+        self.storage = container.resolve(DataStoring.self)
+
+        refreshPushStatus()
+
+        NotificationCenter.default.addObserver(
+            self,
+            selector: #selector(applicationWillEnterForeground),
+            name: UIApplication.willEnterForegroundNotification,
+            object: nil
+        )
+    }
+
+    @objc
+    private func applicationWillEnterForeground(notification: Notification) {
+        refreshPushStatus()
+    }
+
+    func refreshPushStatus(completion: ((UNAuthorizationStatus) -> Void)? = nil) {
+        // Skip call to UNUserNotificationCenter.current() in tests to avoid crashing in package tests
+        #if DEBUG
+        guard ProcessInfo.processInfo.environment["XCTestBundlePath"] == nil else {
+            completion?(pushAuthorizationStatus)
+            return
+        }
+        #endif
+
+        UNUserNotificationCenter.current().getNotificationSettings { [weak self] settings in
+            self?.pushAuthorizationStatus = settings.authorizationStatus
+            completion?(settings.authorizationStatus)
+        }
+    }
+
+    // `completionHandler` should be called iff the function returns true.
+    func didReceiveNotification(userInfo: [AnyHashable: Any], completionHandler: @escaping () -> Void) -> Bool {
+        guard let parsedNotification = ParsedNotification(userInfo: userInfo) else {
+            // Not an Appcues push
+            return false
+        }
+
+        guard let appcues = appcues else {
+            return false
+        }
+
+        // If there's a user ID mismatch, don't do anything with the notification
+        guard parsedNotification.userID == storage.userID else {
+            completionHandler()
+            return true
+        }
+
+        // If no session, start one for the user in the notification
+        if !appcues.isActive {
+            storage.userID = parsedNotification.userID
+            storage.isAnonymous = false
+        }
```

Comment:
im internally questioning this one, seems like it could lead to some weird out of sync issue with the appcues user and the logged in user in the app potentially, in a situation where the app intentionally did not identify that user for some reason - maybe safest to just handle this right alongside L87 - user didnt match so bail out

Reply:
I agree with taking the simpler approach for now. This was provoking all kinds of red flag questions, like "what if the user actually is anonymous?" and "what about identity verification?"